astropy/cosmology/funcs.py

```diff
@@ -109,37 +139,68 @@ def z_at_value(func, fval, zmin=1e-8, zmax=1000, ztol=1e-8, maxfun=500):
+    cosmology, defining a bracket on either side of this maximum will
+    generally return a solution on the same side.
+
+    >>> z_at_value(Planck13.angular_diameter_distance,
+    ...            1500 * u.Mpc, bracket=(1.0, 1.2))  # doctest: +FLOAT_CMP +IGNORE_WARNINGS
+    0.6812769577
+    >>> z_at_value(Planck13.angular_diameter_distance,
+    ...            1500 * u.Mpc, bracket=(2.0, 2.5))  # doctest: +FLOAT_CMP +IGNORE_WARNINGS
+    3.7914913242
+
     Also note that the luminosity distance and distance modulus (two
     other commonly inverted quantities) are monotonic in flat and open
     universes, but not in closed universes.
     """
-    from scipy.optimize import fminbound
+    from scipy.optimize import minimize_scalar
+
+    opt = {'maxiter': maxfun}
+    # Assume custom methods support the same options as default; otherwise user will see warnings.
+    if isinstance(method, str) and method.lower() == 'bounded':
+        opt['xatol'] = ztol
+        if bracket is not None:
+            warnings.warn(f"Option 'bracket' is ignored by method {method}.")
+    else:
+        opt['xtol'] = ztol
 
     fval_zmin = func(zmin)
     fval_zmax = func(zmax)
     if np.sign(fval - fval_zmin) != np.sign(fval_zmax - fval):
-        warnings.warn("""\
-fval is not bracketed by func(zmin) and func(zmax). This means either
-there is no solution, or that there is more than one solution between
-zmin and zmax satisfying fval = func(z).""")
+        warnings.warn(f"fval is not bracketed by func(zmin)={fval_zmin} and func(zmax)="
+                      f"{fval_zmax}. This means either there is no solution, or that there is "
+                      "more than one solution between zmin and zmax satisfying fval = func(z).")
 
     if isinstance(fval_zmin, Quantity):
         val = fval.to_value(fval_zmin.unit)
-        f = lambda z: abs(func(z).value - val)
     else:
-        f = lambda z: abs(func(z) - fval)
+        val = fval
+
+    # 'Brent' and 'Golden' ignore `bounds`, force solution inside zlim
+    def f(z):
+        if z > zmax:
+            return 1.e300 * (1.0 + z - zmax)
+        elif z < zmin:
+            return 1.e300 * (1.0 + zmin - z)
+        else:
+            return abs(Quantity(func(z)).value - val)
 
-    zbest, resval, ierr, ncall = fminbound(f, zmin, zmax, maxfun=maxfun,
-                                           full_output=1, xtol=ztol)
+    res = minimize_scalar(f, method=method, bounds=(zmin, zmax), bracket=bracket, options=opt)
 
-    if ierr != 0:
-        warnings.warn(f'Maximum number of function calls ({ncall}) reached')
+    if not res.success:
+        warnings.warn(f"Solver returned {res.status}: {res.message}\n"
+                      f"Precision {res.fun} reached after {res.nfev} function calls.")
 
-    if np.allclose(zbest, zmax):
-        raise CosmologyError("Best guess z is very close the upper z limit.\n"
+    if verbose:
+        print(res)
```

Comment:
When res.success is False and verbose is True, there is a fair bit of repeat, given the above. But I think that's fine. It's like dumping the trace.

Reply:
Yes, verbose is mostly meant to offer additional info even when successful.
I noted above that 'brent' and 'golden' don't actually include 'status' and 'message', so added a catch for that.

Reply:
This appears to be resolved. 👍